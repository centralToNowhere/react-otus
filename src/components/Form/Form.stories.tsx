import React from "react";
import { Story, Meta } from "@storybook/react";
import { Form, FormProps } from "./Form";

export default {
  title: "Components/PlayerRegistrationForm",
  component: Form,
  args: {
    cellSize: 40,
    capacity: 48,
    speed: 1,
    maxFieldWidth: 600,
    maxFieldHeight: 400,
  },
} as Meta;

const Template: Story<FormProps> = (args) => <Form {...args} />;

export const FormDefault = Template.bind({});
