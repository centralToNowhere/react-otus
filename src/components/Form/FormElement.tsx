import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";

export const FormElement = styled.form`
  display: flex;
  flex-direction: column;
  border: 2px solid ${COLORS.border};
  border-radius: 10px;

  input,
  label,
  p {
    line-height: 1;
  }

  &.game-settings-form {
    .field-error-msg {
      background: ${COLORS.secondary};
    }
  }
`;
