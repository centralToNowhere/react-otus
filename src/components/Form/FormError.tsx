import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";

export const FormError = styled.div`
  color: red;
  padding: 20px;
  border-bottom: 2px solid ${COLORS.border};
`;
