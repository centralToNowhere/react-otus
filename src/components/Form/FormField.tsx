import styled from "@emotion/styled";
import { BREAKPOINTS } from "@/styles/ui-styled";

export const FormField = styled.div`
  display: inline-flex;
  flex-direction: column;
  margin: 20px;

  @media (min-width: ${BREAKPOINTS.sm}) {
    flex: 1;
  }

  input {
    margin: 5px 0 0 0;
  }

  input,
  button {
    min-width: 100px;
  }
`;
