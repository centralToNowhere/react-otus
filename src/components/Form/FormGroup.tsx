import styled from "@emotion/styled";
import { BREAKPOINTS } from "@/styles/ui-styled";

export const FormGroup = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;

export const FormGroupButtons = styled(FormGroup)`
  @media (min-width: ${BREAKPOINTS.sm}) {
    flex-wrap: nowrap;
  }
`;
