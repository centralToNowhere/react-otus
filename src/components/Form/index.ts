export { FormContainer } from "./FormContainer";
export { FormElement } from "./FormElement";
export { FormField } from "./FormField";
export { FormGroup, FormGroupButtons } from "./FormGroup";
export { Form } from "./Form";
export type {
  FormContainerProps,
  IFieldProps,
  IButtonProps,
} from "./FormContainer";
export type { IFormProps } from "./Form";
