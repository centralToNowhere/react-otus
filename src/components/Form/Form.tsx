import React from "react";
import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";
import { minCellSize } from "@/components/Cell";

type ControlCallback<T = void> = (value: T) => void;
type Callback<E, T> = (e: E) => T;
export type FormProps = typeof Form.defaultProps & {
  cellSize?: number;
  onCellSizeChange?: ControlCallback<string>;
  maxFieldWidth?: number;
  onMaxFieldWidthChange?: ControlCallback<string>;
  maxFieldHeight?: number;
  onMaxFieldHeightChange?: ControlCallback<string>;
  readonly cellsInRow?: number;
  readonly cellsInCol?: number;
  capacity?: number;
  onCapacityChange?: ControlCallback<string>;
  speed?: number;
  onSpeedChange?: ControlCallback<string>;
  onStart?: ControlCallback;
  onStop?: ControlCallback;
  onReset?: ControlCallback;
};

type InputProperties =
  | "fieldWidth"
  | "fieldHeight"
  | "cellSize"
  | "capacity"
  | "speed";

type IInputErrors = {
  [key in InputProperties]: {
    show: boolean;
    msg: string;
  };
};

interface IFormState {
  cellSizeString: string;
  capacityString: string;
  maxFieldWidthString: string;
  maxFieldHeightString: string;
  speedString: string;
  errors: IInputErrors;
}

export const isValidNumericString = (
  inputString: unknown
): inputString is number => {
  return (
    typeof inputString === "string" &&
    inputString !== "" &&
    !isNaN(Number(inputString))
  );
};

const isValidCellSizeString = (cellSizeString: unknown): boolean => {
  return (
    isValidNumericString(cellSizeString) &&
    Number(cellSizeString) >= minCellSize
  );
};

export const isValidPositiveNumericString = (str: unknown): boolean => {
  return isValidNumericString(str) && Number(str) > 0;
};

export const isValidNonNegativeNumericString = (str: unknown): boolean => {
  return isValidNumericString(str) && Number(str) >= 0;
};

const floatPattern = "([0-9]*[.])?[0-9]+";
const numberPattern = "[0-9]*";

export class Form extends React.Component<FormProps, IFormState> {
  static defaultProps = {
    onCellSizeChange: (value: string) => {
      console.log(value);
    },
    onMaxFieldWidthChange: (value: string) => {
      console.log(value);
    },
    onMaxFieldHeightChange: (value: string) => {
      console.log(value);
    },
    onCapacityChange: (value: string) => {
      console.log(value);
    },
    onSpeedChange: (value: string) => {
      console.log(value);
    },
    onStart: () => {
      // do nothing
    },
    onStop: () => {
      // do nothing
    },
    onReset: () => {
      // do nothing
    },
  };

  private timer: ReturnType<typeof setTimeout> | null;
  private formRef: React.RefObject<HTMLFormElement>;

  static inputDelay = 1000;

  constructor(props: FormProps) {
    super(props);

    this.state = {
      cellSizeString: String(this.props.cellSize),
      capacityString: String(this.props.capacity),
      maxFieldWidthString: String(this.props.maxFieldWidth),
      maxFieldHeightString: String(this.props.maxFieldHeight),
      speedString: String(this.props.speed),
      errors: {
        fieldWidth: {
          show: false,
          msg: "Expected non-negative number",
        },
        fieldHeight: {
          show: false,
          msg: "Expected non-negative number",
        },
        capacity: {
          show: false,
          msg: "Expected non-negative number",
        },
        cellSize: {
          show: false,
          msg: "Expected number >= 10",
        },
        speed: {
          show: false,
          msg: "Expected positive number",
        },
      },
    };

    this.formRef = React.createRef();
    this.timer = null;
  }

  validateFieldWidth = (fieldWidthString: unknown): boolean => {
    if (isValidNonNegativeNumericString(fieldWidthString)) {
      this.setErrorState("fieldWidth", false);
      return true;
    }

    this.setErrorState("fieldWidth", true);
    return false;
  };

  validateFieldHeight = (fieldHeightString: unknown): boolean => {
    if (isValidNonNegativeNumericString(fieldHeightString)) {
      this.setErrorState("fieldHeight", false);
      return true;
    }

    this.setErrorState("fieldHeight", true);
    return false;
  };

  validateCapacity = (capacityString: unknown): boolean => {
    if (isValidNonNegativeNumericString(capacityString)) {
      this.setErrorState("capacity", false);
      return true;
    }

    this.setErrorState("cellSize", true);
    return false;
  };

  validateCellSize = (cellSizeString: unknown): boolean => {
    if (isValidCellSizeString(cellSizeString)) {
      this.setErrorState("cellSize", false);
      return true;
    }

    this.setErrorState("cellSize", true);
    return false;
  };

  validateSpeed = (speedString: unknown): boolean => {
    if (isValidPositiveNumericString(speedString)) {
      this.setErrorState("speed", false);

      return true;
    }

    this.setErrorState("speed", true);
    return false;
  };

  setErrorState = (property: keyof IInputErrors, show: boolean): void => {
    this.setState((prevState) => {
      return {
        ...prevState,
        errors: {
          ...prevState.errors,
          [property]: {
            ...prevState.errors[property],
            show: show,
          },
        },
      };
    });
  };

  showError = (error: { show: boolean; msg: string }): React.ReactNode => {
    return error.show ? (
      <div className="error-container">
        <p className="error-msg">{error.msg}</p>
      </div>
    ) : null;
  };

  debounce<T>(
    callback: (value: T) => void,
    property: string,
    validator: (value: T) => boolean,
    delay = Form.inputDelay
  ): (value: T) => void {
    return (value) => {
      this.clearTimer();

      this.timer = setTimeout(() => {
        if (validator(value)) {
          callback(value);
          this.setState((prevState) => {
            return {
              ...prevState,
              [property]: Number(value),
            };
          });
        }

        this.timer = null;
      }, delay);
    };
  }

  onChangeFn = (
    onChange: (value: string) => void,
    property: string,
    validator: (value: string) => boolean
  ): Callback<React.ChangeEvent<HTMLInputElement>, void> => {
    const debounced = this.debounce(onChange, property, validator);

    return (e: React.ChangeEvent<HTMLInputElement>) => {
      this.setState((prevState) => {
        return {
          ...prevState,
          [property]: e.target.value,
        };
      });
      debounced(e.target.value);
    };
  };

  onButtonClickFn = <T,>(
    fn: () => T
  ): Callback<React.FormEvent<HTMLButtonElement>, T> => {
    return (e: React.FormEvent<HTMLButtonElement>) => {
      e.preventDefault();
      return fn();
    };
  };

  onBlurFn = <T,>(
    onChange: (value: string) => T,
    validator: (value: string) => T
  ): Callback<React.FocusEvent<HTMLInputElement>, T | void> => {
    return (e: React.FocusEvent<HTMLInputElement>) => {
      this.clearTimer();

      if (validator(e.target.value)) {
        return onChange(e.target.value);
      }
    };
  };

  clearTimer = () => {
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  };

  componentWillUnmount() {
    this.clearTimer();
  }

  render() {
    return (
      <FormElement ref={this.formRef} data-testid={"field-form"}>
        <input autoComplete="off" hidden />
        <FormGroup>
          <FormField>
            <FormLabel htmlFor="field-width">Макс. ширина:</FormLabel>
            <FormInput
              id="field-width"
              type="number"
              pattern={floatPattern}
              step="1"
              name="field-width"
              autoFocus={true}
              value={this.state.maxFieldWidthString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onMaxFieldWidthChange,
                "maxFieldWidthString",
                this.validateFieldWidth
              )}
              onBlur={this.onBlurFn(
                this.props.onMaxFieldWidthChange,
                this.validateFieldWidth
              )}
            />
            {this.showError(this.state.errors.fieldWidth)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="field-height">Макс. высота:</FormLabel>
            <FormInput
              id="field-height"
              type="number"
              pattern={floatPattern}
              step="1"
              name="field-height"
              value={this.state.maxFieldHeightString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onMaxFieldHeightChange,
                "maxFieldHeightString",
                this.validateFieldHeight
              )}
              onBlur={this.onBlurFn(
                this.props.onMaxFieldHeightChange,
                this.validateFieldHeight
              )}
            />
            {this.showError(this.state.errors.fieldHeight)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="cell-size-input">Размер ячейки:</FormLabel>
            <FormInput
              id="cell-size-input"
              type="number"
              pattern={numberPattern}
              step="1"
              name="cell-size-input"
              value={this.state.cellSizeString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onCellSizeChange,
                "cellSizeString",
                this.validateCellSize
              )}
              onBlur={this.onBlurFn(
                this.props.onCellSizeChange,
                this.validateCellSize
              )}
            />
            {this.showError(this.state.errors.cellSize)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="capacity-percentage">
              Процент заполненности:
            </FormLabel>
            <p>{this.state.capacityString} %</p>
            <FormInput
              id="capacity-percentage"
              type="range"
              max="100"
              min="0"
              step="1"
              name="capacity-percentage"
              value={this.state.capacityString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onCapacityChange,
                "capacityString",
                this.validateCapacity
              )}
              onBlur={this.onBlurFn(
                this.props.onCapacityChange,
                this.validateCapacity
              )}
            />
            {this.showError(this.state.errors.capacity)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="speed-change">Обновлений в секунду:</FormLabel>
            <FormInput
              id="speed-change"
              type="number"
              pattern={floatPattern}
              step="0.1"
              name="speed-change"
              value={this.state.speedString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onSpeedChange,
                "speedString",
                this.validateSpeed
              )}
              onBlur={this.onBlurFn(
                this.props.onSpeedChange,
                this.validateSpeed
              )}
            />
            {this.showError(this.state.errors.speed)}
          </FormField>
        </FormGroup>

        <FormGroup>
          <FormField>
            <FormButtonContainer>
              <FormButton
                type="button"
                onClick={this.onButtonClickFn(this.props.onStart)}
                name="start-button"
              >
                Старт
              </FormButton>
            </FormButtonContainer>
          </FormField>

          <FormField>
            <FormButtonContainer>
              <FormButton
                type="button"
                onClick={this.onButtonClickFn(this.props.onStop)}
                name="stop-button"
              >
                Стоп
              </FormButton>
            </FormButtonContainer>
          </FormField>

          <FormField>
            <FormButtonContainer>
              <FormButton
                type="reset"
                onClick={this.onButtonClickFn(this.props.onReset)}
                name="reset-button"
              >
                Сброс
              </FormButton>
            </FormButtonContainer>
          </FormField>
        </FormGroup>
      </FormElement>
    );
  }
}

const FormElement = styled.form`
  margin: 20px;
  display: flex;
  flex-direction: column;
  border: 1px solid black;
  border-radius: 10px;

  input,
  label,
  button,
  p {
    line-height: 1.2;
    font-size: 1.2em;
  }
`;

const FormGroup = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  max-width: 600px;
`;

const FormField = styled.div`
  display: inline-flex;
  flex-direction: column;
  flex: 0 0 calc(100% / 3 - 40px);
  margin: 20px;
  justify-content: end;

  input,
  button {
    margin: 5px 0 0 0;
    min-width: 100px;
  }

  div.error-container {
    position: relative;
  }

  p.error-msg {
    position: absolute;
    line-height: 1em;
    font-size: 1em;
    color: red;
  }
`;

const FormLabel = styled.label`
  display: inline;
`;

const FormInput = styled.input`
  background: inherit;
  border-top: 0;
  border-left: 0;
  border-right: 0;
  display: block;
  outline: 0;
  width: 100px;

  &:focus {
    border-bottom: 2px solid ${COLORS.accent};
  }
`;

const FormButtonContainer = styled.div`
  display: inline-block;
`;

const FormButton = styled.button`
  background: inherit;

  &:focus {
    outline: 2px solid ${COLORS.accent};
    border: 2px solid transparent;
  }
`;
