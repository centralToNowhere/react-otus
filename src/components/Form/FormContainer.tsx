import React, { Dispatch } from "react";
import { AnyAction } from "redux";
import { connect, ConnectedProps } from "react-redux";
import {
  setCellSize,
  setMaxFieldWidth,
  setMaxFieldHeight,
  setCapacity,
  setSpeed,
  FieldMaxWidth,
  FieldMaxHeight,
  FieldCellSize,
  FieldCapacity,
  FieldSpeed,
} from "@/components/Fields";
import { Form } from "@/components/Form";
import { RootState } from "@/store/redux/store";
import { ButtonGameControl } from "@/components/Buttons";
import { l10n } from "@/l10n/ru";
import {
  isAtLeastOneCellDisplayed,
  isValidCellsAmountMax,
  isValidCellSizeString,
  isValidNonNegativeNumericString,
} from "@/utils";

export type Callback<E, T> = (e: E) => T;
export type FormOwnProps = {
  onButtonClickFn?: (e: React.SyntheticEvent<HTMLButtonElement>) => void;
};

export interface IFieldProps<T> {
  formRawData: {
    [Property in keyof T]: T[Property];
  };
  formValidator?: () => boolean;
  onRawChange: Callback<string, void>;
  onChange: Callback<string, void>;
}

export interface IButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  onClick: Callback<React.SyntheticEvent<HTMLButtonElement>, void>;
  content: string;
  disabled?: boolean;
}

const mapStateToProps = (state: RootState) => {
  const { fieldControl, gameField } = state;

  return {
    maxFieldWidth: fieldControl.maxFieldWidth,
    maxFieldHeight: fieldControl.maxFieldHeight,
    capacity: fieldControl.capacity,
    speed: fieldControl.speed,
    cellSize: fieldControl.cellSize,
    gameInProgress: gameField.gameInProgress,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
  setCellSize: (value: string) => {
    dispatch(setCellSize(Number(value)));
  },
  setMaxFieldWidth: (value: string) => {
    dispatch(setMaxFieldWidth(Number(value)));
  },
  setMaxFieldHeight: (value: string) => {
    dispatch(setMaxFieldHeight(Number(value)));
  },
  setCapacity: (value: string) => {
    dispatch(setCapacity(Number(value)));
  },
  setSpeed: (value: string) => {
    dispatch(setSpeed(Number(value)));
  },
});

const connector = connect(mapStateToProps, mapDispatchToProps);
export type FormContainerProps = ConnectedProps<typeof connector> &
  FormOwnProps;

type FormContainerState = {
  rawMaxWidth: string;
  rawMaxHeight: string;
  rawCellSize: string;
  rawSpeed: string;
  rawCapacity: string;
};

class Main extends React.Component<FormContainerProps, FormContainerState> {
  static inputDelay = 1000;

  constructor(props: FormContainerProps) {
    super(props);

    this.state = {
      rawMaxWidth: String(this.props.maxFieldWidth),
      rawMaxHeight: String(this.props.maxFieldHeight),
      rawCellSize: String(this.props.cellSize),
      rawSpeed: String(this.props.speed),
      rawCapacity: String(this.props.capacity),
    };
  }

  setRawData(property: keyof FormContainerState, value: string) {
    this.setState((prevState) => {
      return {
        ...prevState,
        [property]: value,
      };
    });
  }

  setRawMaxWidth = (value: string) => {
    this.setRawData("rawMaxWidth", value);
  };

  setRawMaxHeight = (value: string) => {
    this.setRawData("rawMaxHeight", value);
  };

  setRawCellSize = (value: string) => {
    this.setRawData("rawCellSize", value);
  };

  setRawSpeed = (value: string) => {
    this.setRawData("rawSpeed", value);
  };

  setRawCapacity = (value: string) => {
    this.setRawData("rawCapacity", value);
  };

  getErrorMessage() {
    const cellsMax = isValidCellsAmountMax(
      Number(this.state.rawCellSize),
      Number(this.state.rawMaxWidth),
      Number(this.state.rawMaxHeight)
    );

    const cellsMin = isAtLeastOneCellDisplayed(
      Number(this.state.rawCellSize),
      Number(this.state.rawMaxWidth),
      Number(this.state.rawMaxHeight)
    );

    return cellsMax !== true ? cellsMax : cellsMin !== true ? cellsMin : "";
  }

  getFormError() {
    const msg = this.getErrorMessage();

    return msg ? <p>{msg}</p> : null;
  }

  onChangeRelatedFields = () => {
    this.props.setMaxFieldWidth(this.state.rawMaxWidth);
    this.props.setMaxFieldHeight(this.state.rawMaxHeight);
    this.props.setCellSize(this.state.rawCellSize);
  };

  cellsAmountValidator = () => {
    const cellSize = this.state.rawCellSize;
    const maxHeight = this.state.rawMaxHeight;
    const maxWidth = this.state.rawMaxWidth;

    return (
      isValidNonNegativeNumericString(maxWidth) === true &&
      isValidNonNegativeNumericString(maxHeight) === true &&
      isValidCellSizeString(cellSize) === true &&
      isValidCellsAmountMax(
        Number(cellSize),
        Number(maxWidth),
        Number(maxHeight)
      ) === true &&
      isAtLeastOneCellDisplayed(
        Number(cellSize),
        Number(maxWidth),
        Number(maxHeight)
      ) === true
    );
  };

  render() {
    const {
      onButtonClickFn = () => {
        /** empty **/
      },
    } = this.props;

    return (
      <Form
        fields={
          <>
            <FieldMaxWidth
              formRawData={{
                rawMaxWidth: this.state.rawMaxWidth,
                rawMaxHeight: this.state.rawMaxHeight,
                rawCellSize: this.state.rawCellSize,
                highlight: Boolean(this.getErrorMessage()),
              }}
              formValidator={this.cellsAmountValidator}
              onRawChange={this.setRawMaxWidth}
              onChange={this.onChangeRelatedFields}
            />
            <FieldMaxHeight
              formRawData={{
                rawMaxWidth: this.state.rawMaxWidth,
                rawMaxHeight: this.state.rawMaxHeight,
                rawCellSize: this.state.rawCellSize,
                highlight: Boolean(this.getErrorMessage()),
              }}
              formValidator={this.cellsAmountValidator}
              onRawChange={this.setRawMaxHeight}
              onChange={this.onChangeRelatedFields}
            />
            <FieldCellSize
              formRawData={{
                rawMaxWidth: this.state.rawMaxWidth,
                rawMaxHeight: this.state.rawMaxHeight,
                rawCellSize: this.state.rawCellSize,
                highlight: Boolean(this.getErrorMessage()),
              }}
              formValidator={this.cellsAmountValidator}
              onRawChange={this.setRawCellSize}
              onChange={this.onChangeRelatedFields}
            />
            <FieldCapacity
              formRawData={{
                rawCapacity: this.state.rawCapacity,
              }}
              onRawChange={this.setRawCapacity}
              onChange={this.props.setCapacity}
            />
            <FieldSpeed
              formRawData={{
                rawSpeed: this.state.rawSpeed,
              }}
              onRawChange={this.setRawSpeed}
              onChange={this.props.setSpeed}
            />
          </>
        }
        buttons={
          <>
            <ButtonGameControl
              type={"button"}
              onClick={onButtonClickFn}
              name={"startButton"}
              content={l10n.buttonStart}
              disabled={this.props.gameInProgress}
            />
            <ButtonGameControl
              type={"button"}
              onClick={onButtonClickFn}
              name={"stopButton"}
              content={l10n.buttonStop}
            />
            <ButtonGameControl
              type={"reset"}
              onClick={onButtonClickFn}
              name={"resetButton"}
              content={l10n.buttonReset}
            />
            <ButtonGameControl
              type={"button"}
              onClick={onButtonClickFn}
              name={"generateRandomButton"}
              content={l10n.buttonGenerateRandom}
            />
          </>
        }
        error={this.getFormError()}
      />
    );
  }
}

export const FormContainer = connector(Main);
