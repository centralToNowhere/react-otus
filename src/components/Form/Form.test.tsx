import React from "react";
import userEvent from "@testing-library/user-event/dist";
import { render, screen, waitFor } from "@testing-library/react";
import { Form } from "./Form";
import { l10n } from "@/l10n/ru";

const inputsData = [
  {
    callbackName: "onCellSizeChange",
    labelText: l10n.cellSizeLabel,
    testValues: {
      shouldCallOnChange: ["10", "100", "1000"],
      shouldNotCallOnChange: ["1", "9", "-1", "", "0"],
    },
  },
  {
    callbackName: "onMaxFieldWidthChange",
    labelText: l10n.maxWidthLabel,
    testValues: {
      shouldCallOnChange: ["1", "10", "100", "1000", "0"],
      shouldNotCallOnChange: ["-1", "", "qwerty"],
    },
  },
  {
    callbackName: "onMaxFieldHeightChange",
    labelText: l10n.maxHeightLabel,
    testValues: {
      shouldCallOnChange: ["1", "10", "100", "1000", "0"],
      shouldNotCallOnChange: ["-1", "", "qwerty"],
    },
  },
  {
    callbackName: "onCapacityChange",
    labelText: l10n.capacityLabel,
    testValues: {
      shouldCallOnChange: [
        "1",
        "10",
        "100",
        "1000",
        "0.1",
        "0",
        "-1",
        "qwerty",
      ],
      shouldNotCallOnChange: [""],
    },
  },
  {
    callbackName: "onSpeedChange",
    labelText: l10n.speedLabel,
    testValues: {
      shouldCallOnChange: ["1", "10", "100", "1000", "0.1"],
      shouldNotCallOnChange: ["-1", "0", "", "qwerty"],
    },
  },
];

describe("form tests", () => {
  it("should render form", () => {
    const { asFragment } = render(<Form />);

    const form: HTMLFormElement = screen.getByTestId("field-form");

    expect(form).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render cellSize input", () => {
    render(<Form />);

    const input: HTMLInputElement = screen.getByLabelText(l10n.cellSizeLabel);

    expect(input).toBeInTheDocument();
  });

  it("should render capacity-percentage input", () => {
    render(<Form />);

    const input: HTMLInputElement = screen.getByLabelText(l10n.capacityLabel);

    expect(input).toBeInTheDocument();
  });

  it("should render width input", () => {
    render(<Form />);

    const input: HTMLInputElement = screen.getByLabelText(l10n.maxWidthLabel);

    expect(input).toBeInTheDocument();
  });

  it("should render height input", () => {
    render(<Form />);

    const input: HTMLInputElement = screen.getByLabelText(l10n.maxHeightLabel);

    expect(input).toBeInTheDocument();
  });

  it("should render speed-change", () => {
    render(<Form />);

    const input = screen.getByLabelText(l10n.speedLabel);

    expect(input).toBeInTheDocument();
  });

  it("should render start button", () => {
    render(<Form />);

    const button: HTMLButtonElement = screen.getByText(l10n.buttonStart);

    expect(button).toBeInTheDocument();
  });

  it("should render stop button", () => {
    render(<Form />);

    const button: HTMLButtonElement = screen.getByText(l10n.buttonStop);

    expect(button).toBeInTheDocument();
  });

  it("should render reset button", () => {
    render(<Form />);

    const button: HTMLButtonElement = screen.getByText(l10n.buttonReset);

    expect(button).toBeInTheDocument();
  });
});

describe("input onChange tests", () => {
  const originalInputDelay = Form.inputDelay;

  beforeAll(() => {
    Form.inputDelay = 100;
  });

  afterAll(() => {
    Form.inputDelay = originalInputDelay;
  });

  inputsData.forEach((field) => {
    field.testValues.shouldCallOnChange.forEach((validTestValue) => {
      it(`should call ${field.callbackName} in ${Form.inputDelay} ms for valid input ${validTestValue}`, async () => {
        const callback = jest.fn();

        render(<Form {...{ [field.callbackName]: callback }} />);

        const input: HTMLInputElement = screen.getByLabelText(field.labelText);

        userEvent.clear(input);
        userEvent.type(input, validTestValue);
        await waitFor(
          () => {
            expect(callback).toHaveBeenCalledTimes(1);
          },
          {
            timeout: Form.inputDelay + 100,
          }
        );
      });
    });

    field.testValues.shouldNotCallOnChange.forEach(async (invalidTestValue) => {
      await it(`should not call ${field.callbackName} for invalid input ${invalidTestValue}`, () => {
        const callback = jest.fn();

        render(<Form {...{ [field.callbackName]: callback }} />);
        const input: HTMLInputElement = screen.getByLabelText(field.labelText);

        userEvent.clear(input);
        userEvent.type(input, invalidTestValue);
        return new Promise((resolve, reject) => {
          setTimeout(() => {
            try {
              expect(callback).toHaveBeenCalledTimes(0);
              resolve(true);
            } catch (e) {
              reject(e);
            }
          }, Form.inputDelay + 100);
        });
      });
    });
  });
});

describe("buttons tests", () => {
  it("should run onStart", () => {
    const onStart = jest.fn();

    render(<Form onStart={onStart} />);

    const buttonStart: HTMLButtonElement = screen.getByText(l10n.buttonStart);

    userEvent.click(buttonStart);

    expect(onStart).toHaveBeenCalled();
  });

  it("should run onStop", () => {
    const onStop = jest.fn();

    render(<Form onStop={onStop} />);

    const buttonStop: HTMLButtonElement = screen.getByText(l10n.buttonStop);

    userEvent.click(buttonStop);

    expect(onStop).toHaveBeenCalled();
  });

  it("should run onReset", () => {
    const onReset = jest.fn();

    render(<Form onReset={onReset} />);

    const buttonReset: HTMLButtonElement = screen.getByText(l10n.buttonReset);

    userEvent.click(buttonReset);

    expect(onReset).toHaveBeenCalled();
  });
});
