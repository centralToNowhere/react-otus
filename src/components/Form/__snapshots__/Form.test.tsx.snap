// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form tests should render form 1`] = `
<DocumentFragment>
  <form
    class="css-abmnln-FormElement e1dt6zfi0"
    data-testid="field-form"
  >
    <input
      autocomplete="off"
      hidden=""
    />
    <div
      class="css-1vttlkg-FormGroup e1ftf0n80"
    >
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <label
          class="css-m0dhjx-LabelField e1gymppe0"
          for="field-width"
        >
          Макс. ширина:
        </label>
        <input
          autocomplete="off"
          class="css-iol0j7-InputField e15roiqv0"
          id="field-width"
          name="fieldWidth"
          pattern="([0-9]*[.])?[0-9]+"
          step="1"
          type="number"
          value="undefined"
        />
      </div>
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <label
          class="css-m0dhjx-LabelField e1gymppe0"
          for="field-height"
        >
          Макс. высота:
        </label>
        <input
          autocomplete="off"
          class="css-iol0j7-InputField e15roiqv0"
          id="field-height"
          name="fieldHeight"
          pattern="([0-9]*[.])?[0-9]+"
          step="1"
          type="number"
          value="undefined"
        />
      </div>
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <label
          class="css-m0dhjx-LabelField e1gymppe0"
          for="cell-size"
        >
          Размер ячейки:
        </label>
        <input
          autocomplete="off"
          class="css-iol0j7-InputField e15roiqv0"
          id="cell-size"
          name="cellSize"
          pattern="[0-9]*"
          step="1"
          type="number"
          value="undefined"
        />
      </div>
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <label
          class="css-m0dhjx-LabelField e1gymppe0"
          for="capacity-percentage"
        >
          Процент заполненности:
        </label>
        <p>
          undefined %
        </p>
        <input
          autocomplete="off"
          class="css-1qphvf3-InputField-InputFieldCapacity e1sflu9m0"
          id="capacity-percentage"
          max="100"
          min="0"
          name="capacityPercentage"
          step="1"
          type="range"
          value="undefined"
        />
      </div>
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <label
          class="css-m0dhjx-LabelField e1gymppe0"
          for="speed-change"
        >
          Обновлений в секунду:
        </label>
        <input
          autocomplete="off"
          class="css-iol0j7-InputField e15roiqv0"
          id="speed-change"
          name="speedChange"
          pattern="([0-9]*[.])?[0-9]+"
          step="0.1"
          type="number"
          value="undefined"
        />
      </div>
    </div>
    <div
      class="css-1vttlkg-FormGroup e1ftf0n80"
    >
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <div
          class="css-sabgvl-FormButtonContainer equvaju0"
        >
          <button
            class="css-elonfh-FormButton e2n0d7s0"
            name="startButton"
            type="button"
          >
            Старт
          </button>
        </div>
      </div>
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <div
          class="css-sabgvl-FormButtonContainer equvaju0"
        >
          <button
            class="css-elonfh-FormButton e2n0d7s0"
            name="stopButton"
            type="button"
          >
            Стоп
          </button>
        </div>
      </div>
      <div
        class="css-w6s8l8-FormField eybdsu10"
      >
        <div
          class="css-sabgvl-FormButtonContainer equvaju0"
        >
          <button
            class="css-elonfh-FormButton e2n0d7s0"
            name="resetButton"
            type="reset"
          >
            Сброс
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
